### PR TITLE
osd: collect SMART data from all physical devices

### PR DESCRIPTION
**Pull request by yaarith**
SMART data collection is done by running smartctl via a SubProcessTimed object. It's controlled by two options in common/options.cc: osd_smart_report_interval (default will be 24h) and osd_smart_report_timeout (default is 5s). All devices which start with 'dm-*' are ignored. Still need to handle RAID cases: currently smartctl will be called for each of the devices.
We expect smartctl output to be in a JSON format. Currently the data retrieved is not saved or sent anywhere.

Signed-of-by: Yaarit Hatuka <yaarithatuka@gmail.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
@yaarith instead of put this feature into OSD, an alternative approach is probably to implement a mini daemon using the `rados_service*` API offered by librados, and read the service status in a mgr plugin using `get_daemon_status()` . we can implement all of them in pure python! i am adding the python binding of the `rados_service*` API at #18812.

**Comment by jcsp** [comment]
@tchaikov the downsides to having a separate daemon are that it would need its own connection to the mgr, it would complicate deployment, and it would mean only some systems would have the smart data.  I think there's a strong simplicity argument for building it into the OSD.

File: src/osd/osd_types.cc [hunk 1]
```diff
@@ -4562,6 +4562,7 @@ void OSDSuperblock::encode(bufferlist &bl) const
   ::encode(osd_fsid, bl);
```
Comment (liewegas) [inline src/osd/osd_types.cc]: change the above to ENCODE_START(9, 5, bl)

File: src/osd/osd_types.cc [hunk 2]
```diff
@@ -4595,6 +4596,8 @@ void OSDSuperblock::decode(bufferlist::iterator &bl)
     map<int64_t,epoch_t> pool_last_map_marked_full;
```
Comment (liewegas) [inline src/osd/osd_types.cc]: and DECODE_START about should chagne 8 to 9

Comment (liewegas) [reply src/osd/osd_types.cc]: that will fix the make check error!

File: src/osd/osd_types.cc [hunk 3]
```diff
@@ -4595,6 +4596,8 @@ void OSDSuperblock::decode(bufferlist::iterator &bl)
...
     ::decode(pool_last_map_marked_full, bl);
   }
+  if (struct_v >= 8)
```
Comment (liewegas) [inline src/osd/osd_types.cc]: change this to 9

**Comment by tchaikov** [comment]
> the downsides to having a separate daemon are that it would need its own connection to the mgr, it would complicate deployment,

a separate daemon might complicate deployment. but i don't its own connection to the mgr would.

> and it would mean only some systems (who choose to deploy the extra daemon) would have the smart data. I think there's a strong simplicity argument for building it into the OSD.

yes. because the most likely consumer of the SMART data is a plugin in mgr. and i think, in future, when the mgr plugins are involving faster and their release cycles might be different from that of Ceph, we need to develop and package them in a different model. so instead of embedding this in OSD, would be easier in long term to have a separated tool for collecting this kind of data for their own consumer.

in other words, it's simple to add this to OSD, but from design and maintenance perspective, it's not. IMHO.

**Comment by liewegas** [comment]
I think we should stick this in the OSD until there is a better place to put it.  Part of this effort is associating the devices with OSDs so it is at least possible to correlate it outside the OSD, but we don't have a ceph host agent yet.  Perhaps one will materialize soon!  But until then this will get the job done.

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
@jcsp @tchaikov BTW thinking more about this I'm wondering if this should be refactored as a pull model.  e.g., 'ceph tell osd.123 smart', and let the mgr module grab the data.

The plumbing is relatively simple, here, once smartctl dumps json, so I'm not to concerned with having multiple reporting paths.

File: src/common/options.cc [hunk 4]
```diff
@@ -1606,6 +1606,15 @@ std::vector<Option> get_global_options() {
...
 
+    Option("osd_smart_report_interval", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    //.set_default(86400) // change back when finished!!
+    .set_default(0)
+    .set_description("Frequency (in seconds) for OSD to run smarctl, default is set to 86400"),
+
+    Option("osd_smart_report_timeout", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    .set_default(5)
+    .set_description("Timeout (in seconds) for smarctl to run, default is set to 5"),
+
```
Comment (liewegas) [inline src/common/options.cc]: don't need this
